File: ios/Classes/Bridge/ConsentFlowMethodHandler.swift
//
//  ConsentFlowMethodHandler.swift
//  clever_ads_solutions
//
//  Copyright © 2024 CleverAdsSolutions LTD, CAS.AI. All rights reserved.
//

import CleverAdsSolutions
import Flutter

private let channelName = "cleveradssolutions/consent_flow"

class ConsentFlowMethodHandler: MappedMethodHandler<CASConsentFlow> {
    init(with registrar: FlutterPluginRegistrar) {
        super.init(with: registrar, on: channelName)
    }

    override func initInstance(_ id: String) -> CASConsentFlow {
       return CASConsentFlow()
            .withViewControllerToPresent(Util.findRootViewController())
            .withCompletionHandler(createDismissListener(id))
    }

    override func onMethodCall(_ consentFlow: CASConsentFlow, _ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        switch call.method {
        case "withPrivacyPolicy": withPrivacyPolicy(consentFlow, call, result)
        case "disable": disable(consentFlow, result)
        case "show": show(consentFlow, call, result)
        default: super.onMethodCall(consentFlow, call, result)
        }
    }

    private func withPrivacyPolicy(_ consentFlow: CASConsentFlow, _ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        call.getArgAndReturn("url", result) { url in
            consentFlow.withPrivacyPolicy(url)
        }
    }

    private func disable(_ consentFlow: CASConsentFlow, _ result: @escaping FlutterResult) {
        consentFlow.isEnabled = false
        result(nil)
    }

    private func show(_ consentFlow: CASConsentFlow, _ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        call.getArgAndReturn("force", result) { force in
            if force {
                consentFlow.present()
            } else {
                consentFlow.presentIfRequired()
            }
        }
    }

    private func createDismissListener(_ id: String) -> CASConsentCompletionHandler {
        return { [weak self] status in
            self?.invokeMethod(id, "onDismiss", ["status": status.rawValue])
        }
    }
}

File: android/src/main/kotlin/com/cleveradssolutions/plugin/flutter/bridge/ConsentFlowMethodHandler.kt
package com.cleveradssolutions.plugin.flutter.bridge

import com.cleveradssolutions.plugin.flutter.CASMappedChannel
import com.cleveradssolutions.plugin.flutter.getArgAndReturn
import com.cleveradssolutions.plugin.flutter.success
import com.cleversolutions.ads.ConsentFlow
import io.flutter.embedding.engine.plugins.FlutterPlugin.FlutterPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

private const val CHANNEL_NAME = "cleveradssolutions/consent_flow"

class ConsentFlowMethodHandler(
    binding: FlutterPluginBinding
) : CASMappedChannel<ConsentFlow>(binding, CHANNEL_NAME) {

    override fun initInstance(id: String): ConsentFlow =
        ConsentFlow()
            .withDismissListener(createDismissListener(id))

    override fun onMethodCall(
        instance: ConsentFlow,
        call: MethodCall,
        result: MethodChannel.Result
    ) {
        when (call.method) {
            "withPrivacyPolicy" -> withPrivacyPolicy(instance, call, result)
            "disable" -> disable(instance, result)
            "show" -> show(instance, call, result)
            else -> super.onMethodCall(instance, call, result)
        }
    }

    private fun withPrivacyPolicy(
        consentFlow: ConsentFlow,
        call: MethodCall,
        result: MethodChannel.Result
    ) {
        call.getArgAndReturn<String>("url", result) { url ->
            consentFlow.withPrivacyPolicy(url)
        }
    }

    private fun disable(consentFlow: ConsentFlow, result: MethodChannel.Result) {
        consentFlow.isEnabled = false
        result.success()
    }

    private fun show(consentFlow: ConsentFlow, call: MethodCall, result: MethodChannel.Result) {
        call.getArgAndReturn<Boolean>("force", result) { force ->
            if (force) consentFlow.show() else consentFlow.showIfRequired()
        }
    }

    private fun createDismissListener(id: String): ConsentFlow.OnDismissListener {
        return ConsentFlow.OnDismissListener { status ->
            invokeMethod(id, "onDismiss", hashMapOf("status" to status))
        }
    }

}